flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismAndInputInfosDecider.java

```diff
@@ -75,54 +75,106 @@
+            int initialParallelism,
+            int initialMaxParallelism) {
         checkArgument(
                 initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
                         || initialParallelism > 0);
+        checkArgument(initialMaxParallelism > 0 && initialMaxParallelism >= initialParallelism);
 
         if (consumedResults.isEmpty()) {
             // source job vertex
+            int defaultSourceParallelism = globalDefaultSourceParallelism;
+            if (initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                    && initialMaxParallelism < defaultSourceParallelism) {
+                LOG.info(
+                        "The initial maximum parallelism {} is smaller than the global default source parallelism {}. "
+                                + "Use {} as the final default parallelism of source job vertex {}.",
+                        initialMaxParallelism,
+                        defaultSourceParallelism,
+                        initialMaxParallelism,
+                        jobVertexId);
+                defaultSourceParallelism = initialMaxParallelism;
+            }
+
             int parallelism =
                     initialParallelism > 0 ? initialParallelism : defaultSourceParallelism;
             return new ParallelismAndInputInfos(parallelism, Collections.emptyMap());
-        } else if (initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
-                && areAllInputsAllToAll(consumedResults)
-                && !areAllInputsBroadcast(consumedResults)) {
-            return decideParallelismAndEvenlyDistributeData(
-                    jobVertexId, consumedResults, initialParallelism);
         } else {
-            return decideParallelismAndEvenlyDistributeSubpartitions(
-                    jobVertexId, consumedResults, initialParallelism);
+            int minParallelism = globalMinParallelism;
+            int maxParallelism = globalMaxParallelism;
+
+            if (initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                    && initialMaxParallelism < minParallelism) {
+                LOG.info(
+                        "The initial maximum parallelism {} is smaller than the global minimum parallelism {}. "
+                                + "Use {} as the final minimum parallelism of job vertex {}.",
+                        initialMaxParallelism,
+                        minParallelism,
+                        initialMaxParallelism,
+                        jobVertexId);
+                minParallelism = initialMaxParallelism;
+            }
+            if (initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                    && initialMaxParallelism < maxParallelism) {
+                LOG.info(
+                        "The initial maximum parallelism {} is smaller than the global maximum parallelism {}. "
+                                + "Use {} as the final maximum parallelism of job vertex {}.",
```

Comment:
Maybe "Use {} as the upper bound to decide parallelism of job vertex {}."

Because "maximum parallelism of job vertex" may be misunderstood as `JobVertex#maxParallelism`, however, they are not one thing.

Then, may be also change "the final minimum parallelism" to "the lower bound to decide parallelism"